### PR TITLE
Various improvements for TTreeReader

### PR DESCRIPTION
**Pull request by vepadulano**
See each commit for details

### CODE REVIEW

File: tree/treeplayer/src/TTreeReaderValue.cxx [hunk 1]
```diff
@@ -601,52 +602,35 @@ void ROOT::Internal::TTreeReaderValueBase::CreateProxy() {
...
          }
          if (index == -1) {
-            Error(errPrefix, "The branch %s is contained in a Friend TTree that is not directly attached to the main.\n"
+            Error(errPrefix,
+                  "The branch %s is contained in a Friend TTree that is not directly attached to the main.\n"
                   "This is not yet supported by TTreeReader.",
                   fBranchName.Data());
             return;
          }
-         const char *localBranchName =  originalBranchName.c_str();
```
Comment (pcanal) [inline tree/treeplayer/src/TTreeReaderValue.cxx:625]: Why is this not used?

Comment (vepadulano) [reply tree/treeplayer/src/TTreeReaderValue.cxx:625]: I don't know. It was introduced in [this commit](https://github.com/root-project/root/commit/d35d36eff99d3e9d4881c19e86a98813b743e0c9) and only for TTreeReaderValue, TTreeReaderArray doesn't have that bit. The variable is never referenced anywhere in the function body, nor is this logic used anywhere else in TTreeReader code.

File: tree/treeplayer/src/TTreeReaderArray.cxx [hunk 2]
```diff
@@ -475,12 +475,13 @@ void ROOT::Internal::TTreeReaderArrayBase::CreateProxy()
       if (branch->GetTree() != fTreeReader->GetTree()->GetTree()) {
          // It is in a friend, let's find the 'index' in the list of friend ...
          int index = -1;
-         int current = 0;
-         for(auto fe : TRangeDynCast<TFriendElement>( fTreeReader->GetTree()->GetTree()->GetListOfFriends())) {
-            if (branch->GetTree() == fe->GetTree()) {
-               index = current;
-            }
-            ++current;
+         const auto friendElements =
```
Comment (pcanal) [inline tree/treeplayer/src/TTreeReaderArray.cxx:478]: Why is the new code 'better'? (it seems harder to read/more characters).

Side note (likely in practice irrelevant): the 2 code seems to not do exactly the same things if the same tree is twice in the list of friends.

Comment (vepadulano) [reply tree/treeplayer/src/TTreeReaderArray.cxx:478]: The change uses more modern C++ syntax (and subjectively is more readable to me). I also believe it introduces a fix regarding what you are mentioning: it stops the traversal of the friends list earlier thus is less expensive and avoids picking an eventual second index as the index for this tree.

As a side note, if the situation where the exact same `TTree *` could potentially be used as friend multiple times for the same main tree, then neither the current version nor the previous version of the code would be correct, and that would be a bug that should be addressed separately (outside of this PR)

File: tree/treeplayer/src/TTreeReaderValue.cxx (hunk 1 above)
Comment (pcanal) [reply tree/treeplayer/src/TTreeReaderValue.cxx:625]: From reading that code, it is used in the case there user provided a long name (`friendtreename.branchname`) to access a branch and we want to extract/find the branchname ...

However indeed, the result is not used ...

File: tree/treeplayer/src/TTreeReaderArray.cxx (hunk 2 above)
Comment (pcanal) [reply tree/treeplayer/src/TTreeReaderArray.cxx:478]: The equivalent code in `TTreeReaderValue.cxx:609` has a `break` that make the 2 code (old and std algorithm) strictly equivalent (slight potential difference in number of arithmetic operations).

It is a bit of a matter of taste but I (personally) still find the old code 'clearer' as the "matching criteria" is clearly on its own line:
```
    if (branch->GetTree() == fe->GetTree()) {
```
while in the new code, it is buried in the middle of a 4 lines statements (and at the end of line)
```
         if (auto foundFriend =
                std::find_if(friendElements.begin(), friendElements.end(),
                             [&branch](TFriendElement *fe) { return branch->GetTree() == fe->GetTree(); });
```

**Comment by vepadulano** [comment]
@pcanal I addressed your comments, let me know if this looks good now

**Review by pcanal — APPROVED** [review]
Thanks.

**Comment by vepadulano** [comment]
I will merge these changes together with https://github.com/root-project/root/pull/16165

**Pull request closed without merging**